// instagram-private-api/dist/core/feed.factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const feeds_1 = require("../feeds");
class FeedFactory {
    constructor(client) {
        this.client = client;
    }
    accountFollowers(id) {
        const feed = new feeds_1.AccountFollowersFeed(this.client);
        feed.id = id;
        return feed;
    }
    accountFollowing(id) {
        const feed = new feeds_1.AccountFollowingFeed(this.client);
        feed.id = id;
        return feed;
    }
    directInbox() {
        return new feeds_1.DirectInboxFeed(this.client);
    }
    directThread(options, seqId) {
        const feed = new feeds_1.DirectThreadFeed(this.client);
        feed.id = options.thread_id;
        feed.cursor = options.oldest_cursor;
        feed.seqId = seqId;
        return feed;
    }
    user(id) {
        const feed = new feeds_1.UserFeed(this.client);
        feed.id = id;
        return feed;
    }
    tag(tag) {
        const feed = new feeds_1.TagFeed(this.client);
        feed.tag = tag;
        return feed;
    }
    location(id, tab = 'ranked') {
        const feed = new feeds_1.LocationFeed(this.client);
        feed.id = id;
        feed.tab = tab;
        return feed;
    }
    mediaComments(id) {
        const feed = new feeds_1.MediaCommentsFeed(this.client);
        feed.id = id;
        return feed;
    }
    timeline(reason) {
        const feed = new feeds_1.TimelineFeed(this.client);
        if (reason) {
            feed.reason = reason;
        }
        return feed;
    }
}
exports.FeedFactory = FeedFactory;
